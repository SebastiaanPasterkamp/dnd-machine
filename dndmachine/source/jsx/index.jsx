"use strict";

import React from 'react';
import ReactDom from 'react-dom';
import {BrowserRouter as Router, Switch, Route} from 'react-router-dom';

import LanguageTable from './views/Languages.jsx';
import WeaponsTable from './views/Weapons.jsx';
import DefaultFilter from './views/DefaultFilter.jsx';

let app = document.getElementById('app'),
    filter = document.getElementById('default-filter');

if (app) {
    ReactDom.render(
        <Router>
            <Switch>
                <Route path="/items/languages" component={LanguageTable} />
                <Route path="/items/weapons" component={WeaponsTable} />
            </Switch>
        </Router>,
        app
    );
}

if (filter) {
    ReactDom.render(
        <DefaultFilter/>,
        filter
    );
}