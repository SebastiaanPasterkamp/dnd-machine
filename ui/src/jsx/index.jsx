"use strict";

import '../sass/base.scss';

import React from 'react';
import ReactDom from 'react-dom';
import {BrowserRouter as Router, Switch, Route, Link, Redirect} from 'react-router-dom';

import UiActions from './actions/UiActions.jsx';
import ListDataActions from './actions/ListDataActions.jsx';
import ListDataWrapper from './hocs/ListDataWrapper.jsx';

import AdventureLeagueLogEdit from './views/AdventureLeagueLogEdit.jsx';
import AdventureLeagueLogTable from './views/AdventureLeagueLogTable.jsx';
import AdventureLeagueLogView from './views/AdventureLeagueLogView.jsx';
import ArmorEdit from './views/ArmorEdit.jsx';
import ArmorTable from './views/ArmorTable.jsx';
import CampaignEdit from './views/CampaignEdit.jsx';
import CampaignTable from './views/CampaignTable.jsx';
import CharacterCreate from './views/CharacterCreate.jsx';
import CharacterEdit from './views/CharacterEdit.jsx';
import CharactersTable from './views/CharactersTable.jsx';
import CharactersView from './views/CharactersView.jsx';
import EncounterEdit from './views/EncounterEdit.jsx';
import EncounterTable from './views/EncounterTable.jsx';
import EncounterView from './views/EncounterView.jsx';
import ErrorBoundary from './components/ErrorBoundary.jsx';
import PageFooter from './views/PageFooter.jsx';
import PartyEdit from './views/PartyEdit.jsx';
import PartyTable from './views/PartyTable.jsx';
import PartyView from './views/PartyView.jsx';
import LanguageTable from './views/LanguagesTable.jsx';
import LoadingSplash from './components/LoadingSplash.jsx';
import LoginDialog from './views/LoginDialog.jsx';
import MessageStack from './components/MessageStack.jsx';
import MonsterEdit from './views/MonsterEdit.jsx';
import MonstersTable from './views/MonstersTable.jsx';
import MonsterView from './views/MonsterView.jsx';
import NpcEdit from './views/NpcEdit.jsx';
import NpcTable from './views/NpcTable.jsx';
import NpcView from './views/NpcView.jsx';
import SpellEdit from './views/SpellEdit.jsx';
import SpellView from './views/SpellView.jsx';
import SpellsTable from './views/SpellsTable.jsx';
import UserEdit from './views/UserEdit.jsx';
import UsersTable from './views/UsersTable.jsx';
import UserView from './views/UserView.jsx';
import WeaponEdit from './views/WeaponEdit.jsx';
import WeaponsTable from './views/WeaponsTable.jsx';

import DefaultFilter from './views/DefaultFilter.jsx';
import Navigation from './views/Navigation.jsx';

class DndMachine extends React.Component
{
    render() {
        const { current_user } = this.props;
        const auth = !!current_user;
        const loading = current_user === undefined;

        return <Router><React.Fragment>
        <LoadingSplash
            loading={loading}
            overlay={true}
            />

        <header className="nice-header fixed">
            <div className="nice-header-container">
                <h1 className="nice-header-brand">
                    <Link to={auth ? "/character/list" : "/login"}>
                        <img
                            src="/static/img/dungeons-and-dragons-logo.png"
                            height="35"
                            />
                    </Link>
                </h1>
                <div className="nice-header-collapse">
                    {auth ? <DefaultFilter/> : null }
                    {auth ? <Navigation /> : null }
                </div>

                {auth ? <button
                        type="button"
                        className="nice-header-toggle collapsed"
                        onClick={UiActions.toggleMenu}
                        >
                    <span className="icon-bar top-bar"></span>
                    <span className="icon-bar middle-bar"></span>
                    <span className="icon-bar bottom-bar"></span>
                </button> : null }

            </div>
        </header>

        <section className="nice-fluid-container grid">
            <ErrorBoundary>{ auth ? <Switch>
                <Route
                    path="/items/armor/new"
                    component={ArmorEdit}
                    />
                <Route
                    path="/items/armor/edit/:id"
                    component={ArmorEdit}
                    />
                <Route
                    path="/items/armor/list"
                    component={ArmorTable}
                    />

                <Route
                    path="/campaign/new"
                    component={CampaignEdit}
                    />
                <Route
                    path="/campaign/edit/:id"
                    component={CampaignEdit}
                    />
                <Route
                    path="/campaign/show/:id"
                    render={props => {
                        window.location.href = props.location.pathname;
                        return null;
                    }}
                    />

                <Route
                    path="/campaign/list"
                    component={CampaignTable}
                    />

                <Route
                    path="/character/list"
                    component={CharactersTable}
                    />
                <Route
                    path="/character/show/:id"
                    component={CharactersView}
                    />
                <Route
                    path="/character/edit/:id"
                    component={CharacterEdit}
                    />
                <Route
                    path="/character/new"
                    component={CharacterCreate}
                    />

                <Route
                    path="/log/adventureleague/list"
                    component={AdventureLeagueLogTable}
                    />
                <Route
                    path="/log/adventureleague/show/:id"
                    component={AdventureLeagueLogView}
                    />
                <Route
                    path="/log/adventureleague/edit/:id"
                    component={AdventureLeagueLogEdit}
                    />
                <Route
                    path="/log/adventureleague/new"
                    component={AdventureLeagueLogEdit}
                    />
                <Route
                    path="/log/adventureleague/new/:character_id"
                    component={AdventureLeagueLogEdit}
                    />

                <Route
                    path="/encounter/list"
                    component={EncounterTable}
                    />
                <Route
                    path="/encounter/show/:id"
                    component={EncounterView}
                    />
                <Route
                    path="/encounter/edit/:id"
                    component={EncounterEdit}
                    />
                <Route
                    path="/encounter/new"
                    component={EncounterEdit}
                    />

                <Route
                    path="/monster/list"
                    component={MonstersTable}
                    />
                <Route
                    path="/monster/show/:id"
                    component={MonsterView}
                    />
                <Route
                    path="/monster/edit/:id"
                    component={MonsterEdit}
                    />
                <Route
                    path="/monster/new"
                    component={MonsterEdit}
                    />

                <Route
                    path="/party/list"
                    component={PartyTable}
                    />
                <Route
                    path="/party/show/:id"
                    component={PartyView}
                    />
                <Route
                    path="/party/edit/:id"
                    component={PartyEdit}
                    />
                <Route
                    path="/party/new"
                    component={PartyEdit}
                    />

                <Route
                    path="/items/languages/list"
                    component={LanguageTable}
                    />

                <Route
                    path="/items/spell/list"
                    component={SpellsTable}
                    />
                <Route
                    path="/items/spell/show/:id"
                    component={SpellView}
                    />
                <Route
                    path="/items/spell/new"
                    component={SpellEdit}
                    />
                <Route
                    path="/items/spell/edit/:id"
                    component={SpellEdit}
                    />

                <Route
                    path="/items/weapon/new"
                    component={WeaponEdit}
                    />
                <Route
                    path="/items/weapon/edit/:id"
                    component={WeaponEdit}
                    />
                <Route
                    path="/items/weapon/list"
                    component={WeaponsTable}
                    />

                <Route
                    path="/npc/list"
                    component={NpcTable}
                    />
                <Route
                    path="/npc/new"
                    component={NpcEdit}
                    />
                <Route
                    path="/npc/edit/:id"
                    component={NpcEdit}
                    />
                <Route
                    path="/npc/show/:id"
                    component={NpcView}
                    />

                <Route
                    path="/user/list"
                    component={UsersTable}
                    />
                <Route
                    path="/user/new"
                    component={UserEdit}
                    />
                <Route
                    path="/user/edit/:id"
                    component={UserEdit}
                    />
                <Route
                    path="/user/show/:id"
                    component={UserView}
                    />

                <Redirect
                    path="/login"
                    to="/character/list"
                    />

                <Route path="/logout" render={props => {
                    ListDataActions.doLogout(
                        () => props.history.push('/login')
                    );

                    return null;
                }} />

                <Route path="*" render={props => {
                    return "Whoops";
                }} />
            </Switch> : <Switch>

                <Route path="*" render={props => (
                    <LoginDialog
                        message="Welcome to D&amp;D Machine. Please log in using your credentials to access this website."
                        icon="d20"
                        />
                )} />

            </Switch> }</ErrorBoundary>
        </section>
        <PageFooter />
        <MessageStack
            messages={[
                {id: 'abc', message: 'test', title: 'foo'}
                ]}
            />
    </React.Fragment></Router>;
    }
}

const App = ListDataWrapper(DndMachine, ['current_user']);

ReactDom.render(
    <App />,
    document.getElementById('app')
);
