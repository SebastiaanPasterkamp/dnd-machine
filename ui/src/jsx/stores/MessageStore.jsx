import React from 'react';
import Reflux from 'reflux';
import _ from 'lodash';

import ReportingActions from '../actions/ReportingActions.jsx';

export class MessageStore extends Reflux.Store
{
    constructor()
    {
        super();
        this.state = {
            messages: []
        };
        this.listenables = ReportingActions;
    }

    onShowMessage(type, message, title=null, timeout=0)
    {
        const msg = {
            id: _.uniqueId('message_'),
            type,
            message,
            title,
            timeout,
        };
        const messages = _.concat(this.state.messages, [msg]);
        this.setState({messages});

        ReportingActions.showMessage.completed();

        if (!msg.timeout) {
            return;
        }
        _.delay(
            ReportingActions.hideMessage,
            msg.timeout * 1000,
            msg.id
        );
    }

    onHideMessage(id)
    {
        const messages = _.reject(
            this.state.messages,
            {id}
        );
        this.setState({messages});

        ReportingActions.hideMessage.completed();
    }
}

export default Reflux.initStore(MessageStore);