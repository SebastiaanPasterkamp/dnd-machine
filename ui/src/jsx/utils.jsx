import {
    entries,
    filter,
    forEach,
    flow,
    get,
    intersection,
    isArray,
    keys,
    reduce,
    replace,
} from 'lodash/fp';
import { create } from 'mathjs'
const math = create();
// math.import(require('mathjs/lib/expression/function/eval'));

// create simple functions for all operators
math.import({
    // arithmetic
    add:        function (a, b) { return a + b },
    subtract:   function (a, b) { return a - b },
    multiply:   function (a, b) { return a * b },
    divide:     function (a, b) { return a / b },
    mod:        function (a, b) { return a % b },
    unaryPlus:  function (a) { return  a },
    unaryMinus: function (a) { return -a },

    // bitwise
    bitOr:           function (a, b) { return a | b },
    bitXor:          function (a, b) { return a ^ b },
    bitAnd:          function (a, b) { return a & b },
    bitNot:          function (a) { return ~a },
    leftShift:       function (a, b) { return a << b },
    rightArithShift: function (a, b) { return a >> b },
    rightLogShift:   function (a, b) { return a >>> b },

    // logical
    or:  function (a, b) { return !!(a || b) },
    xor: function (a, b) { return !!a !== !!b },
    and: function (a, b) { return !!(a && b) },
    not: function (a) { return !a },

    // relational
    equal:     function (a, b) { return a == b },
    unequal:   function (a, b) { return a != b },
    smaller:   function (a, b) { return a < b },
    larger:    function (a, b) { return a > b },
    smallerEq: function (a, b) { return a <= b },
    largerEq:  function (a, b) { return a >= b },
    pi: Math.PI,
    e: Math.E,
    'true': true,
    'false': false,
    'null': null,
});

var allFromMath = {};
Object.getOwnPropertyNames(Math).forEach(function (name) {
    if (!Object.prototype.hasOwnProperty(name)) {
        allFromMath[name] = Math[name];
    }
});
math.import(allFromMath);

const styleCache = {
    '': null,
};

const utils = {
    randomColor() {
        var letters = '0123456789ABCDEF';
        var color = '#';
        for (var i = 0; i < 6; i++) {
            color += letters[Math.floor(Math.random() * 16)];
        }
        return color;
    },

    makeStyle(conditionals, styles=[]) {
        const active = filter(
            null,
            reduce(
                (active, style) => {
                    if (conditionals[style]) {
                        active.push(style);
                    }
                    return active;
                },
                styles
            )(keys(conditionals))
        );
        const style = active.join(' ');
        if (styleCache[style] === undefined) {
            styleCache[style] = style;
        }
        return styleCache[style];
    },

    closest(options, target, defaultOption=null) {
        let matched = reduce(
            (match, option) => {
                const delta = Math.abs(options[option] - target);
                if (delta < match.delta) {
                    return { delta, option };
                }
                return match;
            },
            { option: defaultOption, delta: Number.MAX_SAFE_INTEGER }
        )(keys(options));
        return matched.option;
    },

    resolveMath(obj, formula, name='') {
        const replacings = {};
        const re = new RegExp(/\b[a-z_]+\.[a-z_.]+\b/g);
        const prefix = new RegExp(`^${name}\.`);

        forEach((match) => {
            const path = replace(prefix, '', match);
            const value = get(path, obj);
            if (value !== undefined || path !== match) {
                replacings[match] = JSON.stringify(value);
            }
        })(formula.match(re));

        flow(entries, forEach(([match, value]) => {
            formula = replace(match, value, formula);
        }))(replacings);

        try {
            const result = math.eval(formula);
            if (result.toArray !== undefined) {
                return result.toArray();
            }
            return result;
        } catch(err) {
            return undefined;
        }
    }
};

export function userHasRole(user, role) {
    const roles = isArray(role) ? role : [role];
    return (
        intersection(
            user.role || [],
            roles
        ).length > 0
    );
};

export function memoize(name, callback) {
    if (!('_memoize' in this)) {
        this._memoize = {};
    }
    if (!(name in this._memoize)) {
        this._memoize[name] = callback;
    }
    return this._memoize[name];
}

export default utils;
