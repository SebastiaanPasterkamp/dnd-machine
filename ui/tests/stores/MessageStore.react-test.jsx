import React from 'react';
import Reflux from 'reflux';
import _ from 'lodash';
jest.useFakeTimers();

import {MessageStore} from '../../src/jsx/stores/MessageStore.jsx';
jest.mock('../../src/jsx/actions/ReportingActions.jsx');

describe('MessageStore', () => {
    var store;

    beforeEach(() => {
        Reflux.initStore(MessageStore);
        store = new MessageStore();

        _.uniqueId = jest.fn();
        _.uniqueId
            .mockReturnValueOnce('id_1')
            .mockReturnValueOnce('unexpected_2');
    });

    it('capture a message', function() {

        store.onShowMessage(
            'good', 'Hello world'
        );
        jest.runAllTimers();
        expect(store.state).toMatchSnapshot();
    });

    it('remove a message', function() {

        store.onShowMessage(
            'muted', 'Goodbye world', 'Bye', 10
        );
        jest.runAllTimers();
        expect(store.state).toMatchSnapshot();

        store.onHideMessage("id_1")
        jest.runAllTimers();
        expect(store.state).toMatchSnapshot();
    });
});
