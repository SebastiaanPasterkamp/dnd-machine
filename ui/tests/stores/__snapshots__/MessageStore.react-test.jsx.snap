// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageStore capture a message 1`] = `
Object {
  "messages": Array [
    Object {
      "id": "id_1",
      "message": "Hello world",
      "timeout": 0,
      "title": null,
      "type": "good",
    },
  ],
}
`;

exports[`MessageStore remove a message 1`] = `
Object {
  "messages": Array [
    Object {
      "id": "id_1",
      "message": "Goodbye world",
      "timeout": 10,
      "title": "Bye",
      "type": "muted",
    },
  ],
}
`;

exports[`MessageStore remove a message 2`] = `
Object {
  "messages": Array [],
}
`;
